source/opt/loop_descriptor.h

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_LOOP_DESCRIPTORS_H_
+#define LIBSPIRV_OPT_LOOP_DESCRIPTORS_H_
+
+#include <algorithm>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "opt/module.h"
+#include "opt/pass.h"
+#include "opt/tree_iterator.h"
+
+namespace spvtools {
+namespace ir {
+class CFG;
+class LoopDescriptor;
+
+// A class to represent and manipulate a loop in structured control flow.
+class Loop {
+  // The type used to represent nested child loops.
+  using ChildrenList = std::vector<Loop*>;
+
+ public:
+  using iterator = ChildrenList::iterator;
+  using const_iterator = ChildrenList::const_iterator;
+  using BasicBlockListTy = std::unordered_set<uint32_t>;
+
+  Loop()
+      : loop_header_(nullptr),
+        loop_continue_(nullptr),
+        loop_merge_(nullptr),
+        loop_preheader_(nullptr),
+        parent_(nullptr) {}
+
+  Loop(IRContext* context, opt::DominatorAnalysis* analysis, BasicBlock* header,
+       BasicBlock* continue_target, BasicBlock* merge_target);
+
+  // Iterators which allows access to the nested loops.
```

Comment:
Not clear which loops will be visited.  I get the impression it will only be the immediately nested loops (or some other appropriate language).  In a triply nested loop, the begin and end will not visit the inner most loop.  Is that correct?

Reply:
Yes, this allow only the visit of the children in the tree. That's required for to allow DF tree iterators to work properly. The iterator of LoopDescriptor do a post-order visit of the tree.

But we could add post_begin/post_end and pre_begin/pre_end visitors, it is kind of trivial now.

Reply:
Could you update the comment to make it clearer?  I was not looking for a change in behaviour, but a more precise comment.

Reply:
I updated it